org.eclipse.lsp4e/src/org/eclipse/lsp4e/VersionedResult.java

```diff
@@ -0,0 +1,45 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Cocotec Ltd and others.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Ahmed Hussain (Cocotec Ltd) - initial implementation
+ *
+ *******************************************************************************/
+package org.eclipse.lsp4e;
+
+/**
+ * Groups a result from the language server with the modification stamp that existed on the document
+ * at the time the request to the server was made
+ * @param <T>
+ */
+public class VersionedResult<T> {
```

Comment:
I don't think it's worth a new API. IMO, this should be verified internally by the particular operation that requires such checks (eg made internal to `LSPFormatter`).

Reply:
That's what this API is designed to do. It doesn't oblige consuming code to take version conflicts into account, it just allows them to do so.

The only thing that knows what version of the document the server will have seen when a request (e.g. a format request) is sent, is the `DocumentContentSynchronizer` for that document. We need a wrapper object like this to bundle the version and the result together to make the version check atomic and consistent. Otherwise client code would have to make two separate calls to `DocumentContentSynchronizer`.

If `T` is some edits, then when we schedule those on the event thread, we can read the document version (knowing it can only change on the event thread) and compare it to `VersionedResult.getStamp()`. If they're different, then the results are stale and the client code can either abandon the changes, apply them anyway (!), or restart the operation (the classic way of handling a write consistency clash).

Reply:
If we have some `getUpToDateServer` as discussed above, would we still need such an API or would it be good enough?

Reply:
Unfortunately you do need this kind of API to guarantee correctness. Without blocking the UI thread you can't ever guarantee that the server is up to date, only that it is consistent with all the changes it has received _so far_. Hence the need for an API which is effectively `computeBasedOnWhatYouKnowAndLetMeFigureOutWhetherItIsStale`.